Dynamic compliance tests

yisraelU opened the pull request:
[@Baccata's notes]

* Changes the default timestamp format used in Json/Documents to be `EPOCH_SECOND`, in an effort to align with the semantics of the AWS protocols. This is **behaviour breaking**, but is for the better as it aligns the behaviour of what Smithy4s does with the behaviour of the AWS-owned smithy libraries/code-generators, and reduces the hoops we have to jump through so that Smithy4s-provided tooling can talk to AWS-provided tooling. It is a hard decision to make, as we've tried to avoid breaking behaviour in the past, but is very much the correct decision for the sanity of the maintainers, and the users who read the Smithy documentation. 

* Amends the compliance tests module to make it easier to run tests from a dynamically loaded module. This is gonna let us pull new tests as they come by virtue of bumping versions, without having to inspect the test specifications to know what we need to add.  

* Wires a bunch of cherry-picked tests from the Aws Rest Json protocol. This allows us to assert how much work there is and what are the differences between our "SimpleRestJson" protocol and the official "AWS Rest Json" protocol. Eventually, we'd like to support the latter out of the box, but devil's in the detail and it's a slow process.

yisraelU committed:
adds model transformer to deal with protocol change from restJson1 to simpleRestJson
and removed Malformed Operations

yisraelU committed:
Merge branch 'series/0.17' into dynamic-complianceTests

# Conflicts:
#	build.sbt

yisraelU committed:
add transformers module

yisraelU committed:
comment out malformed operations transformer

Baccata commented on modules/aws-http4s/test/src-ce3/smithy4s/compliancetests/ProtocolComplianceTest.scala:
```diff
@@ -0,0 +1,154 @@
+          .clientAndServerTests(SimpleRestJsonIntegration, wrapper.service)
+      })
+  }
+  val whitelist = Set(
```
```suggestion
  val allowlist = Set(
```

Baccata commented on modules/aws-http4s/test/src-ce3/smithy4s/compliancetests/ProtocolComplianceTest.scala:
```diff
@@ -0,0 +1,154 @@
+  private def loadDynamic(
+                           doc: Document
+                         ): Either[PayloadError, DynamicSchemaIndex] = {
+    Document.decode[Model](doc).map(_.filter(whitelist)).map(load)
```
```suggestion
    Document.decode[Model](doc).map(_.filter(allowlist)).map(load)
```

Baccata commented on modules/transformers/src/ProtocolTransformer.scala:
```diff
@@ -0,0 +1,23 @@
+      )
+      map
+    }
+    ctx.getTransformer().renameShapes(ctx.getModel(), rename)
```
I think there's a `mapTrait` that would be better suited : you're not supposed to rename the protocol, but change the protocol the service knows about

yisraelU replied:
ok good to know

yisraelU replied:
actually quite annoying as the HttpRequestTestCase are not traits rather Nodes that contain the protocol

yisraelU committed:
more config

yisraelU committed:
Merge branch 'series/0.17' into dynamic-complianceTests

yisraelU committed:
removed MalformedOperationsTransformer.scala
Add support for pulling config list from Alloy and creating AllowList
update dumpModel to cache per output file path

yisraelU committed:
update model transformer to use maptrait

yisraelU committed:
remove fs2 circe dependency

Baccata commented on modules/http4s/test/src-ce3/smithy4s/compliancetests/ProtocolComplianceTest.scala:
```diff
@@ -131,22 +150,69 @@ object ProtocolComplianceTest extends EffectSuite[IO] with BaseCatsSuite {
+        IO(
+          new java.io.BufferedReader(
+            new java.io.InputStreamReader(
+              getClass.getResourceAsStream("/META-INF/json/test-config.json")
```
That method is JVM specific : you'll have to dump it onto the file system at build time.

Alternatively, we could put this configuration into a smithy-compliant json file as metadata, so that it'd already benefit from the build logic you added to dump the model as json file. 

We'd then add a `def metadata: Map[String, Document]` to our `DynamicSchemaIndex`, that could be really useful in the future.

yisraelU replied:
yes , I am aware , thats why its still in draft , i'm just trying to dot a couple i's before tackling this

yisraelU replied:
I actually had a plan to move this to Alloy , so the dumping logic , decoding can live in one place , to provide reuse for the Disney addition too

yisraelU replied:
done

yisraelU committed:
remove fs2 circe dependency

yisraelU committed:
Merge branch 'series/0.17' into dynamic-complianceTests

yisraelU committed:
update allow rule to use metadata from smithy model

yisraelU committed:
fix for metadata encode

yisraelU committed:
fix for metadata decode

yisraelU committed:
Merge branch 'metaencode-queryParams-fix' into metacodec-fix

yisraelU committed:
merging both encode and decode

yisraelU committed:
Merge branch 'metacodec-fix' into dynamic-complianceTests

yisraelU committed:
update metadata name

yisraelU committed:
modify schema of fake impl

yisraelU committed:
fmt

yisraelU committed:
Merge branch 'series/0.17' into dynamic-complianceTests

yisraelU committed:
netter handling of empty test case bodies

yisraelU committed:
Adds MappedHintsSchemaVisitor
a bunch of other fixes

yisraelU committed:
update alloy to 0.1.14

yisraelU committed:
smithy specific document decoder

yisraelU committed:
smithy specific document decoder

Baccata committed:
Fix tests in Scala 2.12

yisraelU committed:
update alt

yisraelU committed:
headers added

yisraelU committed:
Merge branch 'series/0.17' into dynamic-complianceTests

yisraelU committed:
fmt

yisraelU committed:
update docs

Baccata commented on modules/compliance-tests/src/smithy4s/compliancetests/internals/MappedHintsSchemaVisitor.scala:
```diff
@@ -0,0 +1,144 @@
+import smithy4s.compliancetests.internals.MappedHintsSchemaVisitor.hintMask
+import smithy4s.schema.Schema._
+
+object MappedHintsSchemaVisitor {
```
I think we could do this more concisely. Remind me the problem this solves again ?

Baccata replied:
Alright, I remember ... I'm started to think the better course of action would be to align on what AWS does, and change the behaviour for 0.18. 

The maintenance burden induced by the difference is starting to be somewhat uncomfortable. I think aligning would be preferable, for our own sake, as well as for the sake of the ecosystem, considering the difference of default violates the principle of least-surprise ... 

In order to reduce the risk based on the change of behaviour in user's applications, we could add a smithy-level validator (in alloy) that would issue a warning for any timestamp shape connected transitively to a service with `alloy#simpleRestJson` stating that it's preferable for users to be explicit about timestamp formats ... 

What do you think ?

Baccata replied:
@yisraelU, I've created the `series/0.18`. When you have the time, could you : 

- [x] Change this PR to target that new series branch
- [x] Sync this PR with the new target 
- [x] Change the default timestamp format to EPOCH_SECONDS in the relevant Json/Document schema visitors 
- [x] Remove the then un-necessary Service/Endpoint transformation 

Thank you in advance ;)

yisraelU committed:
add host as input to test

yisraelU committed:
Merge branch 'series/0.18' into dynamic-complianceTests

yisraelU committed:
change default timestamp in SchemaVisitorJCodec and remove unnecessary code

yisraelU committed:
modified defaults to Epoch_Seconds for Document codecs and Opts

yisraelU committed:
update OptsVisitorSpec with new default

yisraelU committed:
update docs

Baccata committed:
Merge remote-tracking branch 'origin/series/0.18' into dynamic-complianceTests

Baccata committed:
Avoid reliance on sys.error

Baccata committed:
Simplify the decoding of canonical smithy representation

This requires a fix to the Document Decoder logic

Baccata committed:
cleanup
